[BEAM-3008] Adds parameters templatization for Bigtable

dmytroivanov4206 opened the pull request:
* Adds ValueProvider for instanceId, projectId and tableId
* Removes BigtableOptions deprecated methods
* Updates existing unit tests

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

sduskis commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -240,63 +233,62 @@ static Read create() {
      */
-    @Deprecated
-    public Read withBigtableOptions(BigtableOptions.Builder optionsBuilder) {
+    public Read withTableId(ValueProvider<String> tableId) {
```
Unfortunately, we need to keep public versions of `withBigtableOptions()` and `getBigtableOptions()`.  We need to maintain backwards compatible APIs

sduskis commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -149,10 +151,10 @@
-   * {@link BigtableIO.Read#withBigtableOptions(BigtableOptions) BigtableOptions} that specifies
-   * the source Cloud Bigtable instance, and a {@link BigtableIO.Read#withTableId tableId} that
-   * specifies which table to read. A {@link RowFilter} may also optionally be specified using
+   * initialized with a {@link BigtableIO.Read#withInstanceId(ValueProvider)} and
```
The default usage will be the String values rather than ValueProviders for ease of use and brevity of code.  Can you please use `{@link BigtableIO.Read#withInstanceId(String)}` and etc in the comments?

dmytroivanov4206 replied:
Ok.

dmytroivanov4206 commented:
@sduskis  I have update PR to address your comments, please take a look, thank you.

sduskis commented:
There are transient failures in other connectors.  @lukecwik can you please review and commit this change?

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -240,23 +243,64 @@ static Read create() {
      *
```
You can call any withInstanceId method so you shouldn't say that they need to call the string one.

Please update this here and elsewhere, ditto for withProjectId, ...

dmytroivanov4206 replied:
ok

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java:
```diff
@@ -176,7 +176,7 @@ public void testReadBuildsCorrectly() {
     BigtableIO.Read read = BigtableIO.read().withBigtableOptions(BIGTABLE_OPTIONS);
 
-    thrown.expect(IllegalArgumentException.class);
+    thrown.expect(NullPointerException.class);
```
This should really still continue to throw the IllegalArgumentException

How do you know the correct exception is being thrown since NPE/IllegalArgument could refer to many incorrect parameters?

dmytroivanov4206 replied:
ok

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -1119,7 +1185,7 @@ public BigtableWriteException(KV<ByteString, Iterable<Mutation>> record, Throwab
   static void validateTableExists(BigtableConfig config, PipelineOptions options) {
     if (config.getValidate()) {
-      String tableId = config.getTableId();
+      String tableId = config.getTableId().get();
```
Does it make sense to skip validation if any of the value providers are inaccessible?

Otherwise users will always need to write withTable(myValueProvider).withoutValidation();

dmytroivanov4206 replied:
As per documents, this should be called right before pipeline run, should not be providers value available at that point?

lukecwik replied:
Value providers are only useful during pipeline execution and add no value during pipeline construction. If you know the value during pipeline construction why would you need a value provider?

What documents do you refer to so they can be updated/fixed?

dmytroivanov4206 replied:
I was referring to PTransform#validate JavaDoc.

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTest.java:
```diff
@@ -0,0 +1,238 @@
+    assertEquals(INSTANCE_ID.get(), config.withInstanceId(INSTANCE_ID).getInstanceId().get());
+
+    thrown.expect(IllegalArgumentException.class);
+    config.withProjectId(null);
```
withInstanceId?

dmytroivanov4206 replied:
thanks

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTest.java:
```diff
@@ -0,0 +1,238 @@
+    assertEquals(TABLE_ID.get(), config.withTableId(TABLE_ID).getTableId().get());
+
+    thrown.expect(IllegalArgumentException.class);
+    config.withProjectId(null);
```
withTableId?

dmytroivanov4206 replied:
thanks

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTest.java:
```diff
@@ -0,0 +1,238 @@
+    assertEquals(SERVICE, config.withBigtableService(SERVICE).getBigtableService());
+
+    thrown.expect(IllegalArgumentException.class);
+    config.withProjectId(null);
```
withBigtableService?

dmytroivanov4206 replied:
thanks

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    assertEquals(PROJECT_ID.get(), service.getBigtableOptions().getProjectId());
+    assertEquals(INSTANCE_ID.get(), service.getBigtableOptions().getInstanceId());
+    assertEquals(CredentialOptions.CredentialType.SuppliedCredentials,
```
This check can only run on machines that have Google application default credentials configured.

dmytroivanov4206 replied:
Addressed

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTest.java:
```diff
@@ -0,0 +1,238 @@
+    assertEquals(PROJECT_ID.get(), service.getBigtableOptions().getProjectId());
+    assertEquals(INSTANCE_ID.get(), service.getBigtableOptions().getInstanceId());
+    assertEquals(true, service.getBigtableOptions().getBulkOptions().useBulkApi());
+    assertEquals(CredentialOptions.CredentialType.SuppliedCredentials,
```
This can't run on a machine that doesn't have Google application default credentials configured in their environment.

dmytroivanov4206 replied:
Addressed

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -588,7 +654,7 @@ Write withBigtableService(BigtableService bigtableService) {
       getBigtableConfig().validate();
 
-      input.apply(ParDo.of(new BigtableWriterFn(getBigtableConfig().getTableId(),
+      input.apply(ParDo.of(new BigtableWriterFn(getBigtableConfig().getTableId().get(),
```
This will be resolved at pipeline construction time effectively making it so that static value providers can only be used.

dmytroivanov4206 replied:
Addressed

dmytroivanov4206 commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfig.java:
```diff
@@ -249,11 +242,11 @@ public String toString() {
 
     if (getProjectId() != null) {
-      effectiveOptions.setProjectId(getProjectId());
+      effectiveOptions.setProjectId(getProjectId().get());
```
Addressed

dmytroivanov4206 commented:
@lukecwik  I have addressed your comments, please take a look. The only thing is validation  on read and write level,  documentation says that it will be called right before running pipeline so providers data should be available at that point, or I am wrong?

lukecwik commented:
Note that a pipeline template in Dataflow is created by running the pipeline with the TemplatingDataflowRunner which saves the job description in Dataflow. The validation happens while still executing on the users machine before the template is created.

ValueProviders are only useful during pipeline execution.

dmytroivanov4206 commented:
In this case if we have tableId provided via dynamic value provider there is no way to validate if table is there. I can modify validate to only check it for static value provider. Do you think this is a way to go?

dmytroivanov4206 commented:
@lukecwik Please take a look again on this PR, I have fixed everything you mentioned except for validation (hopefully will finish this on Monday).

dmytroivanov4206 commented:
@sduskis what do you think if we validate if table exist only in case bigtable options were provided or value provider has accessible value, in other cases we will not do it?

sduskis commented:
@dmytroivanov4206: Given the inherent constraint in templates and validation, we don't have a choice in the matter.  I'm fine with limiting the validation to non-template scenarios.  Templates will fail at runtime if the table does not exist.

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfig.java:
```diff
@@ -179,12 +179,8 @@ void populateDisplayData(DisplayData.Builder builder) {
     if (getBigtableOptions() != null) {
       builder.add(DisplayData.item("bigtableOptions", getBigtableOptions().toString())
-        .withLabel("Bigtable Options"));
+        .withLabel("Provided Bigtable Options"));
```
Are you trying to differentiate options given by the user and the "effective" options?

dmytroivanov4206 replied:
Not anymore, before we had those here. Fixed.

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -1061,7 +1042,7 @@ public RowFilter getRowFilter() {
     }
 
     public String getTableId() {
```
Even though this method is public, the class it is in isn't.

We should update this method to return the ValueProvider<String> and have callers deal with it instead of returning null or drop this method and migrate all current callers to `config.getTableId()`

dmytroivanov4206 replied:
Done

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java:
```diff
@@ -1185,7 +1166,7 @@ public BigtableWriteException(KV<ByteString, Iterable<Mutation>> record, Throwab
   static void validateTableExists(BigtableConfig config, PipelineOptions options) {
     if (config.getValidate()) {
-      String tableId = config.getTableId().get();
+      String tableId = checkNotNull(config.getTableId().get());
```
.get() will throw an exception if its not accessible. You need to check that its accessible and skip the check if it isn't.

dmytroivanov4206 replied:
Done.

lukecwik commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfig.java:
```diff
@@ -151,47 +151,36 @@ BigtableConfig withValidate(boolean isEnabled) {
 
   void validate() {
-    checkArgument(getProjectId() != null && !getProjectId().isEmpty()
+    checkArgument(getTableId() != null, "Could not obtain Bigtable table id");
```
If the table id is accessible, check that it isn't empty.

dmytroivanov4206 replied:
Done

lukecwik commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java:
```diff
@@ -166,21 +165,12 @@ public void testReadBuildsCorrectly() {
 
-  @Test
-  public void testReadValidationFailsMissingTable() {
-    BigtableIO.Read read = BigtableIO.read().withBigtableOptions(BIGTABLE_OPTIONS);
```
This test should still work if the user supplies `""` as the table.

dmytroivanov4206 replied:
After changing validate, yes.

lukecwik commented:
I'm with @sduskis, skip validation if a required parameter for validation is missing. Also add tests to show that validation is being skipped if the parameters are from value providers that are not accessible.

dmytroivanov4206 commented:
@sduskis , @lukecwik  I have updated code to reflect your comments, please take a look and I will add more tests to the BigtableConfig/IOTest afterwards.

dmytroivanov4206 commented:
@sduskis , @lukecwik  please take a look on this PR, I have added the changes you have requested and requested unit tests.

sduskis commented:
@chamikaramj, what else do we need to do at this point?
